### ISSUE TITLE
Feature idea: instantiate a PackedScene with arguments

### ISSUE THREAD

**Issue by ttencate** [Cat-Lips/GodotSharp.SourceGenerators#26]
It's common to have a (C#) script on the root node of a scene, and wanting to instantiate that scene through code. Unfortunately, `PackedScene.Instantiate()` always invokes the default (no-args) constructor, so it doesn't allow us to send any arguments for initialization.

A workaround is to have a two-stage initialization approach, which typically looks like this:

```csharp
var scene = ResourceLoader.Load<PackedScene>("res://...");
var myNode = scene.Instantiate<MyNode>();
myNode.Prop1 = value1;
myNode.Prop2 = value2;
```

However, this suffers from some drawbacks:

1. It's easy to forget to set some required property. We could make this less likely by providing a `MyNode.Init(value1, value2)` method, but even so it's still possible to forget to call that method.
2. Even if we don't forget anything, the compiler doesn't know that, so it'll throw warnings on non-nullable fields if `<Nullable>enable</Nullable>` is enabled in the `.csproj` file.

My boilerplate to work around both of these issues:

```csharp
class MyNode
{
    public Type1 Prop1;
    public Type2 Prop2;

    public void Init(Type1 prop1, Type2 prop2)
    {
        Prop1 = prop1;
        Prop2 = prop2;
    }

    // BEGIN BOILERPLATE

    #pragma warning disable CS8618
    public MyNode()
    {
    }
    #pragma warning restore CS8618

    public static MyNode Instantiate(Type1 prop1, Type2 prop2)
    {
        var scene = ResourceLoader.Load<PackedScene>("res://...");
        var myNode = scene.Instantiate<MyNode>();
        myNode.Init(prop1, prop2);
        return myNode;
    }

    // END BOILERPLATE
}
```

Now I can simply call it in one line, safely:

```csharp
var my_node = MyNode.Instantiate(value1, value2);
```

Adding the no-arg constructor which suppresses [CS8618](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/nullable-warnings#nonnullable-reference-not-initialized) makes the warnings on individual non-nullable fields go away. Possibly this constructor could even be `private`; I haven't tried.

Anyway, would it be possible to add a source generator to this (already awesome!) library to get rid of the boilerplate? It seems most of the pieces are already there, but I'm not well versed enough in source generator land to put them together.

**Comment by Cat-Lips** [comment]
I like the cut o' yer jib.  I'll see what I can do :)

**Comment by Cat-Lips** [comment]
Hey tt.  I've attached a zipped nuget package to the pull request.  That should allow you to do the following:
```
[OnInstantiate]
private void Initialise(Type1 prop1, Type2 prop2)
{
    ...
}
```
to generate:
```
public static MyNode Instantiate(Type1 prop1, Type2 prop2)
{
    ...
}
```
Let me know how it goes and we'll merge it in :)

PS:  Haven't addressed the CS8618 issue (but can if we need)
PPS:  Haven't provided a way to customise tscn path (but can if we need)

**Comment by ttencate** [comment]
Wow, awesome, thanks a lot! I'll try to find some time to test it this weekend.

I think it would make sense to generate a `private` no-arg constructor as I suggested, to make it harder to accidentally use `new MyScene()` instead of `MyScene.Instantiate()` and skip the required initialization. I tried this in my own hacky source generator and it seems that Godot doesn't mind that the ctor is private.

I noticed that you didn't copy `using` declarations over from the original file, but just put some common ones in ([here](https://github.com/Cat-Lips/GodotSharp.SourceGenerators/pull/27/files#diff-5ca6f1b059d2187b52d1bf5bc00874cdeb5976df2fd9dc36899e8716293781a3)). That could lead to confusing errors if an argument type is not in one of these namespaces. It's well beyond my source-generator-fu to fix that, though.

**Comment by Cat-Lips** [comment]
I've just added some tests to ensure that anything generated will use a full namespace in the declaration, so that should be ok as is.

I think it would be ok to generate a default (protected) constructor if none already exists, so long as no other generator has the same idea!

**Comment by Cat-Lips** [comment]
Try this :)

[GodotSharp.SourceGenerators.2.1.0-230617-1625.Release.nupkg.zip](https://github.com/Cat-Lips/GodotSharp.SourceGenerators/files/11778145/GodotSharp.SourceGenerators.2.1.0-230617-1625.Release.nupkg.zip)

**Comment by Cat-Lips** [comment]
New pre-release package with requested changes uploaded to nuget.  Will formalise into 2.1 release in a few days to give early adopters time to provide feedback (just in case).

**Comment by ttencate** [comment]
Sorry for the lack of respect. Been very busy, and I kept thinking "I'll get round to it today, no need to send a message", but then didn't...

**Comment by ttencate** [comment]
Hmm, this is weird. VS Code (with OmniSharp) thinks everything is fine; I can Ctrl+click into the generated `Instantiate` method and see its implementation. However, both `dotnet build` and the Build button in Godot give me an error at the call site:

```
$ dotnet build
MSBuild version 17.4.4+8e9cb2952 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
CSC : warning CS9057: The analyzer assembly '/home/thomas/.nuget/packages/godotsharp.sourcegenerators/2.1.0/analyzers/dotnet/cs/GodotSharp.SourceGenerators.dll' references version '4.5.0.0' of the compiler, which is newer than the currently running version '4.4.0.0'. [/tmp/repro/Repro.csproj]
/tmp/repro/Main.cs(15,25): error CS0117: 'Main' does not contain a definition for 'Instantiate' [/tmp/repro/Repro.csproj]

Build FAILED.

CSC : warning CS9057: The analyzer assembly '/home/thomas/.nuget/packages/godotsharp.sourcegenerators/2.1.0/analyzers/dotnet/cs/GodotSharp.SourceGenerators.dll' references version '4.5.0.0' of the compiler, which is newer than the currently running version '4.4.0.0'. [/tmp/repro/Repro.csproj]
/tmp/repro/Main.cs(15,25): error CS0117: 'Main' does not contain a definition for 'Instantiate' [/tmp/repro/Repro.csproj]
    1 Warning(s)
    1 Error(s)

Time Elapsed 00:00:01.07
```

Minimal reproduction project: [repro.zip](https://github.com/Cat-Lips/GodotSharp.SourceGenerators/files/11856734/repro.zip)

I'm using Godot 4.1-beta3 from the official builds, on Arch Linux with dotnet SDK 7.0.7 installed:

```
$ dotnet --version
7.0.107
```

I'm not sure where the warning about compiler version is coming from; maybe you compiled with a .NET 8 prerelease?

Any idea what I'm doing wrong?

**Comment by Cat-Lips** [comment]
Ok, ta.

With Visual Studio on Windows, I get:
- dotnet --version = 7.0.304
- csc.exe -version = 4.6.0-3.23259.8 (c3cc1d0c)

But the package reference for Microsoft.CodeAnalysis.CSharp is v4.5.0, so could possibly be that.

Here's a version compiled with CodeAnalysis v4.4.0.  Let me know if that fixes it.

[GodotSharp.SourceGenerators.2.1.0.nupkg.zip](https://github.com/Cat-Lips/GodotSharp.SourceGenerators/files/11858890/GodotSharp.SourceGenerators.2.1.0.nupkg.zip)

**Comment by ttencate** [comment]
Yes, that does fix it!

For the record, I added `NuGet.Config` with the following contents:

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="local-packages" value="thirdparty/packages" />
  </packageSources>
</configuration>
```

Then I put the package (unzipped) into `thirdparty/packages/GodotSharp.SourceGenerators.2.1.0.nupkg` and installed it using `dotnet add package GodotSharp.SourceGenerators`. I did remove `~/.nuget/` at some point to avoid caching the official release, but I'm not sure that was necessary.

It's annoying that the version mismatch only comes up as a warning, even though it causes errors down the line (not even sure how or why). So to use the official releases, people just need to use the latest version of .NET? I am new to this C# and .NET world, so I'm afraid I can't be much help here...

**Comment by ttencate** [comment]
Unrelated: it seems like `?` in parameter types does not work.

```csharp
public partial class Item : MarginContainer
{
    [OnInstantiate]
    private void Init(int? units) { }
}
```

This gives:

> CSC : warning CS8785: Generator 'OnInstantiateSourceGenerator' failed to generate source. It will not contribute to the output and compilation errors may occur as a result. Exception was of type 'ArgumentException' with message 'The hintName 'Item.Init(int?).g.cs' contains an invalid character '?' at position 30. (Parameter 'hintName')'

Interestingly, using `Nullable<int>` does not work either.

**Comment by Cat-Lips** [comment]
Yep, NuGet.Config is the way to do it.  Glad it fixed it.  At least we know the cause.
(Note that my editor build fails when using local nuget packages - It registers the paths, but fails to check after nuget.org fails.)

So unfortunately, yes, it looks like consumers may need to compile using the latest .NET7, but can still target .NET6, etc.  (I expect your csc.exe equivalent would return version 4.4)

Alternatively, I can downgrade the analyser package and do a point release.

**Comment by Cat-Lips** [comment]
Regarding nullable issue.  Linux has a different set of InvalidFileNameChars to windows, but the generator internals does its own check for invalid chars.  I'll have to copy the windows output to use for all platforms.

**Comment by Cat-Lips** [comment]
2.2.1 should fix both those issues :)

**Comment by ttencate** [comment]
Great work, thank you!

**Comment by ttencate** [comment]
Works like a charm!
